tritium-lib/src/main/java/com/palantir/tritium/proxy/ByteBuddyInstrumentation.java

```diff
@@ -0,0 +1,234 @@
+import net.bytebuddy.implementation.FieldAccessor;
+import net.bytebuddy.implementation.LoadedTypeInitializer;
+import net.bytebuddy.implementation.MethodCall;
+import net.bytebuddy.implementation.MethodDelegation;
+import net.bytebuddy.implementation.bytecode.assign.Assigner;
+import net.bytebuddy.jar.asm.Opcodes;
+import net.bytebuddy.matcher.ElementMatchers;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+final class ByteBuddyInstrumentation {
+
+    private static final Logger log = LoggerFactory.getLogger(ByteBuddyInstrumentation.class);
+    // Offset to avoid duplicate fqcns
+    private static final AtomicInteger offset = new AtomicInteger();
+    // Reuse generated classes when possible
+    private static final TypeCache<List<Class<?>>> cache = new TypeCache.WithInlineExpunction<>(TypeCache.Sort.WEAK);
+    private static final Joiner UNDERSCORE_JOINER = Joiner.on('_');
+
+    private ByteBuddyInstrumentation() {
+        throw new UnsupportedOperationException();
+    }
+
+    @SuppressWarnings("unchecked")
+    static <T, U extends T> T instrument(
+            Class<T> interfaceClass,
+            U delegate,
+            List<InvocationEventHandler<InvocationContext>> handlers,
+            InstrumentationFilter instrumentationFilter) {
+        checkNotNull(interfaceClass, "interfaceClass");
+        checkNotNull(delegate, "delegate");
+        checkNotNull(instrumentationFilter, "instrumentationFilter");
+        checkNotNull(handlers, "handlers");
+
+        if (!isAccessible(interfaceClass)) {
+            log.warn("Interface {} is not accessible. Delegate {} of type {} will not be instrumented",
+                    SafeArg.of("interface", interfaceClass),
+                    UnsafeArg.of("delegate", delegate),
+                    SafeArg.of("delegateType", delegate.getClass()));
+            return delegate;
+        }
+
+        // Use the interface classloader to avoid creating additional proxies for other loaders delegating
+        // to the same type.
+        ClassLoader classLoader = getClassLoader(interfaceClass);
+        ImmutableList<Class<?>> interfaces = Arrays.stream(Proxies.interfaces(interfaceClass, delegate.getClass()))
+                .filter(iface -> isAccessibleFrom(classLoader, iface))
+                .collect(ImmutableList.toImmutableList());
+        try {
+            return (T) newInstrumentationClass(classLoader, interfaces)
+                    .getConstructor(interfaceClass, InvocationEventHandler.class, InstrumentationFilter.class)
+                    .newInstance(delegate, CompositeInvocationEventHandler.of(handlers), instrumentationFilter);
+        } catch (ReflectiveOperationException e) {
+            throw new SafeRuntimeException("Failed to instrumented delegate", e);
+        }
+    }
+
+    private static Class<?> newInstrumentationClass(ClassLoader classLoader, List<Class<?>> interfaces) {
+        Class<?> interfaceClass = interfaces.get(0);
```

Comment:
This assumes primary interface is first, do we test this?

Reply:
I went back and forth on passing `ifaceClass, additionalInterfaces` vs a single list of interfaces, this approach lead to fewer lines, but your raise a good point that it comes at expense to clarity, I'll refactor.

Reply:
I don't think you necessarily have to refactor, I'd just suggest docs + test explaining assumption

Reply:
I read a blog post a few years ago that I can't seem to find anymore, but the gist was that we should push failures/validation to the "left" in the chain of compiler -> unit tests -> integration/ete -> runtime. I think the refactor I've pushed obviates the need for a unit test.

Reply:
I'm fully supportive of having the compiler robots identify issues earlier in the pipeline 😄